### PR TITLE
CUAEV backward

### PR DESCRIPTION
**Pull request by yueyericardo**
Try to implement backward for cuaev.

### CODE REVIEW

File: torchani/cuaev/aev.cu [hunk 1]
```diff
@@ -1,24 +1,32 @@
...
+#include <vector>
 
 #include <ATen/Context.h>
 #include <THC/THC.h>
 #include <c10/cuda/CUDACachingAllocator.h>
 #include <THC/THCThrustAllocator.cuh>
 
 #define PI 3.141592653589793
+typedef torch::Tensor Tensor;
+typedef torch::autograd::tensor_list tensor_list;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply torchani/cuaev/aev.cu:14]: How about also

```C++
using torch::IValue
```

File: torchani/cuaev/aev.cu [hunk 2]
```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
...
+    auto species_t = saved[1];
+    auto tensor_Rij = saved[2];
+    auto tensor_radialRij = saved[3];
+    auto tensor_angularRij = saved[4];
+    auto EtaR_t = saved[5], ShfR_t = saved[6], EtaA_t = saved[7], Zeta_t = saved[8], ShfA_t = saved[9],
+         ShfZ_t = saved[10];
+
+    c10::intrusive_ptr<c10::ivalue::Tuple> aev_params_tp_ptr = ctx->saved_data["aev_params"].toTuple();
+    auto aev_params_tp = aev_params_tp_ptr->elements();
+    AEVScalarParams<float> aev_params = {(float)aev_params_tp[0].toDouble(),
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:792]: How about creating a ctor

```C++
AEVScalarParams(const IValue &iv)
```
Then you can do
```C++
AEVScalarParams<float> aev_params = ctx->saved_data["aev_params"];
```

File: torchani/cuaev/aev.cu [hunk 3]
```diff
@@ -406,19 +497,30 @@ void initConsts(AEVScalarParams<float>& aev_params, cudaStream_t stream) {
...
+  int nAngularRij;
+} Result;
+
 // NOTE: assumes size of EtaA_t = Zeta_t = EtaR_t = 1
 template <typename ScalarRealT = float>
-torch::Tensor cuComputeAEV(
-    torch::Tensor coordinates_t,
-    torch::Tensor species_t,
+Result cuaev_forward(
+    Tensor& coordinates_t,
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:514]: Could you make it `const Tensor &`? Same for other arguments.

File: torchani/cuaev/aev.cu [hunk 4]
```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
         align<4>(aev_params.angular_length),
         ncenter_atoms);
   }
-  return aev_t;
+  return {
+      aev_t, aev_params, tensor_Rij, tensor_radialRij, tensor_angularRij, total_natom_pairs, nRadialRij, nAngularRij};
+}
+
+Tensor cuaev_backward(
+    Tensor& grad_output,
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:687]: Could you make it `const Tensor &`? Same for other arguments.

File: torchani/cuaev/aev.cu [hunk 5]
```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
...
+      grad_radial_dist.packed_accessor32<float, 1, torch::RestrictPtrTraits>(),
+      grad_coord.packed_accessor32<float, 3, torch::RestrictPtrTraits>(),
+      d_radialRij,
+      nRadialRij);
+
+  return grad_coord;
+}
+
+#define AEV_INPUT                                                                                                     \
+  Tensor &coordinates_t, Tensor &species_t, double Rcr_, double Rca_, Tensor &EtaR_t, Tensor &ShfR_t, Tensor &EtaA_t, \
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:742]: Could you make it `const Tensor &`? Same for other arguments.

File: torchani/cuaev/aev.cu [hunk 6]
```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
...
+                              res.tensor_angularRij,
+                              EtaR_t,
+                              ShfR_t,
+                              EtaA_t,
+                              Zeta_t,
+                              ShfA_t,
+                              ShfZ_t});
+      ctx->saved_data["aev_params"] = res.aev_params.toTuple();
+      c10::List<int64_t> int_list({res.total_natom_pairs, res.nRadialRij, res.nAngularRij});
+      ctx->saved_data["int_list"] = torch::IValue(int_list);
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:774]: Can we consense this into
```C++
ctx->saved_data["int_list"] = c10::List<int64_t>{res.total_natom_pairs, res.nRadialRij, res.nAngularRij};
```

File: torchani/cuaev/aev.cu [hunk 7]
```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
...
-  m.def("cuComputeAEV", &cuComputeAEV<float>);
+  m.def("cuComputeAEV", cuaev_cuda);
+}
+
+TORCH_LIBRARY_IMPL(cuaev, CUDA, m) {
+  m.impl("cuComputeAEV", cuaev_cuda);
+}
+
+TORCH_LIBRARY_IMPL(cuaev, Autograd, m) {
+  m.impl("cuComputeAEV", cuaev_autograd);
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:842]: Does this work?
```C++
m.impl("cuComputeAEV", CuaevAutograd::apply);
```

**Comment by zasdfgbnm** [comment]
Reopen to trigger some code review tools

File: torchani/cuaev/aev.cu (hunk 7 above)
Comment (yueyericardo) [reply torchani/cuaev/aev.cu:842]: The official dispatcher example is using the same style, so I guess it's good? https://github.com/pytorch/tutorials/blob/master/advanced_source/dispatcher/op.cpp#L77

File: torchani/cuaev/aev.cu [hunk 8]
```diff
@@ -574,11 +694,154 @@ torch::Tensor cuComputeAEV(
...
+                              res.tensor_Rij,
+                              res.tensor_radialRij,
+                              res.tensor_angularRij,
+                              EtaR_t,
+                              ShfR_t,
+                              EtaA_t,
+                              Zeta_t,
+                              ShfA_t,
+                              ShfZ_t});
+      ctx->saved_data["aev_params"] = static_cast<torch::IValue>(res.aev_params);
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:787]: Do you need this explicit cast? Will implicit casting work?
```C++
ctx->saved_data["aev_params"] = res.aev_params;
```

Comment (yueyericardo) [reply torchani/cuaev/aev.cu]: solved

File: torchani/cuaev/aev.cu [hunk 9]
```diff
@@ -105,6 +128,42 @@ __global__ void pairwiseDistance(
   }
 }
 
+// every block compute blocksize RIJ's gradient by column major, to avoid atomicAdd waiting
+template <typename DataT, typename IndexT = int>
+__global__ void pairwiseDistance_backward(
+    torch::PackedTensorAccessor32<DataT, 3, torch::RestrictPtrTraits> pos_t,
+    torch::PackedTensorAccessor32<DataT, 1, torch::RestrictPtrTraits> grad_radial_dist,
+    torch::PackedTensorAccessor32<DataT, 3, torch::RestrictPtrTraits> grad_coord,
+    PairDist<DataT>* d_radialRij,
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:137]:
```suggestion
    const PairDist<DataT>* d_radialRij,
```

File: torchani/cuaev/aev.cu [hunk 10]
```diff
@@ -287,6 +346,53 @@ __global__ void cuRadialAEVs(
...
 
+// every <THREADS_PER_RIJ> threads take care of 1 RIJ, and iterate <nShfR / THREADS_PER_RIJ> times
+template <typename SpeciesT, typename DataT, int THREADS_PER_RIJ>
+__global__ void cuRadialAEVs_backward(
+    torch::PackedTensorAccessor32<SpeciesT, 2, torch::RestrictPtrTraits> species_t,
+    torch::PackedTensorAccessor32<DataT, 1, torch::RestrictPtrTraits> ShfR_t,
+    torch::PackedTensorAccessor32<DataT, 1, torch::RestrictPtrTraits> EtaR_t,
+    torch::PackedTensorAccessor32<DataT, 3, torch::RestrictPtrTraits> grad_output,
+    torch::PackedTensorAccessor32<DataT, 1, torch::RestrictPtrTraits> grad_radial_dist,
+    PairDist<DataT>* d_Rij,
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:357]:
```suggestion
    const PairDist<DataT>* d_Rij,
```

**Comment by yueyericardo** [comment]
Torchani backward:
```
2/2 [========] - 1s 388ms/step - rmse: 20.8459
   GPU Memory Cached (pytorch) :  4912.0MB / 8119.6MB (GeForce GTX 1080)
   GPU Memory Used (nvidia-smi):  5609.5MB / 8119.6MB (GeForce GTX 1080)
=> More detail about benchmark PER EPOCH
   Total AEV - 85.1 ms
   Forward - 13.3 ms
   Backward - 481.5 ms
   Force - 183.6 ms
   Optimizer - 9.2 ms
   Others - 3.9 ms
   Epoch time - 776.7 ms
```
Initial cuaev backward:
```
2/2 [========] - 1s 663ms/step - rmse: 1323.6314
   GPU Memory Cached (pytorch) :  1634.0MB / 8119.6MB (GeForce GTX 1080)
   GPU Memory Used (nvidia-smi):  2331.5MB / 8119.6MB (GeForce GTX 1080)
=> More detail about benchmark PER EPOCH
   Total AEV - 22.7 ms
   Forward - 13.9 ms
   Backward - 637.2 ms
   Force - 639.9 ms
   Optimizer - 8.8 ms
   Others - 3.8 ms
   Epoch time - 1.326 sec
```

**Comment by yueyericardo** [comment]
After using share memory to avoid AtomicAdd
```
share_mem needed: 23840
2/2 [========] - 1s 310ms/step - rmse: 1323.6313
   GPU Memory Cached (pytorch) :  1634.0MB / 8119.6MB (GeForce GTX 1080)
   GPU Memory Used (nvidia-smi):  2331.5MB / 8119.6MB (GeForce GTX 1080)
=> More detail about benchmark PER EPOCH
   Total AEV - 22.7 ms
   Forward - 13.9 ms
   Backward - 285.4 ms
   Force - 286.3 ms
   Optimizer - 9.1 ms
   Others - 3.8 ms
   Epoch time - 621.2 ms
```

**Comment by yueyericardo** [comment]
After optimize some `pow()` function
```
2/2 [========] - 0s 133ms/step - rmse: 1323.6313
   GPU Memory Cached (pytorch) :  1434.0MB / 8119.6MB (GeForce GTX 1080)
   GPU Memory Used (nvidia-smi):  2131.5MB / 8119.6MB (GeForce GTX 1080)
=> More detail about benchmark PER EPOCH
   Total AEV - 23.4 ms
   Forward - 13.9 ms
   Backward - 105.6 ms
   Force - 106.4 ms
   Optimizer - 13.2 ms
   Others - 4.0 ms
   Epoch time - 266.6 ms
```

**Comment by yueyericardo** [comment]
Remove a lot of (warpsize * nbr) unnecessary shared mem, use warplevel aggregate to accumulate grad.
```
2/2 [========] - 0s 102ms/step - rmse: 1323.6313
   GPU Memory Cached (pytorch) :  1434.0MB / 8119.6MB (GeForce GTX 1080)
   GPU Memory Used (nvidia-smi):  2131.5MB / 8119.6MB (GeForce GTX 1080)
=> More detail about benchmark PER EPOCH
   Total AEV - 23.2 ms
   Forward - 14.0 ms
   Backward - 76.7 ms
   Force - 76.5 ms
   Optimizer - 9.6 ms
   Others - 3.8 ms
   Epoch time - 203.8 ms
```
Note that the backward timing include network backward also.
Some timing from nsight
```
cuAngularAEVs
Begins: 1.57992s
Ends: 1.58472s (+4.805 ms)
grid:  <<<19585, 1, 1>>>
block: <<<64, 1, 1>>>
Launch Type: Regular
Static Shared Memory: 0 bytes
Dynamic Shared Memory: 4,128 bytes
Registers Per Thread: 60
Local Memory Per Thread: 0 bytes
Local Memory Total: 54,394,880 bytes
Shared Memory executed: 69,632 bytes
Shared Memory Bank Size: 4 B
Theoretical occupancy: 50 %
Launched from thread: 16615
Latency: ←9.781 μs
Correlation ID: 8404
Stream: Default stream (7)
```

```
cuAngularAEVs_backward
Begins: 1.61448s
Ends: 1.62205s (+7.564 ms)
grid:  <<<19585, 1, 1>>>
block: <<<64, 1, 1>>>
Launch Type: Regular
Static Shared Memory: 0 bytes
Dynamic Shared Memory: 2,560 bytes
Registers Per Thread: 80
Local Memory Per Thread: 0 bytes
Local Memory Total: 55,705,600 bytes
Shared Memory executed: 30,720 bytes
Shared Memory Bank Size: 4 B
Theoretical occupancy: 37.5 %
Launched from thread: 16752
Latency: ←7.508 ms
Correlation ID: 11068
Stream: Default stream (7)
```

The issue now is `Registers Per Thread: 80` exceeds the limit 64.

File: torchani/cuaev/aev.cu [hunk 11]
```diff
@@ -10,7 +10,7 @@
 #include <THC/THCThrustAllocator.cuh>
 
 #define PI 3.141592653589793
-#define WRAPSIZE 32
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu]: You can use `warpSize` from CUDA

File: torchani/cuaev/aev.cu [hunk 12]
```diff
@@ -1,24 +1,48 @@
+#include <stdio.h>
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:1]: Remove this

File: torchani/cuaev/aev.cu [hunk 13]
```diff
@@ -1,24 +1,48 @@
...
 #include <cub/cub.cuh>
+#include <vector>
 
 #include <ATen/Context.h>
 #include <THC/THC.h>
 #include <c10/cuda/CUDACachingAllocator.h>
 #include <THC/THCThrustAllocator.cuh>
 
 #define PI 3.141592653589793
+#define WARPSIZE 32
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:13]: Don't define your own warp size. Use `warpSize` from CUDA or `C10_WARP_SIZE` from pytorch instead.

File: torchani/cuaev/aev.cu [hunk 14]
```diff
@@ -222,7 +282,7 @@ __global__ void cuAngularAEVs(
         DataT fc_ijk = fc_ij * fc_ik;
 
         IndexT subaev_offset = csubaev_offsets[type_j * num_species + type_k];
-        IndexT aev_offset = aev_params.radial_length + subaev_offset;
+        // IndexT aev_offset = aev_params.radial_length + subaev_offset;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:285]: Remove unused line

File: torchani/cuaev/aev.cu [hunk 15]
```diff
@@ -247,6 +307,258 @@ __global__ void cuAngularAEVs(
...
+    PairDist<DataT>* d_centralAtom,
+    int* d_nPairsPerCenterAtom,
+    int* d_centerAtomStartIdx,
+    AEVScalarParams<DataT, IndexT> aev_params,
+    int maxnbrs_per_atom_aligned,
+    int angular_length_aligned,
+    int ncentral_atoms) {
+  extern __shared__ DataT smem[];
+
+  int threads_per_catom = TILEX * TILEY;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:332]: Could you do
```C++
constexpr int threads_per_catom = TILEX * TILEY;
static_assert(threads_per_catom == C10_WARP_SIZE);
```

File: torchani/cuaev/aev.cu [hunk 16]
```diff
@@ -247,6 +307,258 @@ __global__ void cuAngularAEVs(
...
+  for (int jj = laneIdx; jj < jnum; jj += threads_per_catom) {
+    sdjx_grad[jj] = 0;
+    sdjy_grad[jj] = 0;
+    sdjz_grad[jj] = 0;
+  }
+
+  short2 tile = make_short2(laneIdx % TILEX, laneIdx / TILEX);
+  const DataT tc = 0.95; // theta constant factor
+  // must sync if threads_per_catom != 32 (wrap size)
+  // __syncthreads
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:425]: Could you add the same comment to forward as well?

File: torchani/cuaev/aev.cu [hunk 17]
```diff
@@ -273,7 +585,7 @@ __global__ void cuRadialAEVs(
   int i = d.i;
   int j = d.j;
 
-  SpeciesT type_i = species_t[mol_idx][i];
+  // SpeciesT type_i = species_t[mol_idx][i];
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:588]: Remove unused line

File: torchani/cuaev/aev.cu [hunk 18]
```diff
@@ -406,19 +765,36 @@ void initConsts(AEVScalarParams<float>& aev_params, cudaStream_t stream) {
   delete[] subaev_offsets;
 }
 
+typedef struct Result {
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:768]: `typedef` is only needed in C. So please remove it on C++.

File: torchani/cuaev/aev.cu [hunk 19]
```diff
@@ -463,8 +839,9 @@ torch::Tensor cuComputeAEV(
 
   // buffer to store all the pairwise distance (Rij)
   auto total_natom_pairs = n_molecules * max_natoms_per_mol * max_natoms_per_mol;
-  auto buffer_Rij = allocator.allocate(sizeof(PairDist<float>) * total_natom_pairs);
-  PairDist<float>* d_Rij = (PairDist<float>*)buffer_Rij.get();
+  auto d_options = torch::dtype(torch::kUInt8).device(torch::kCUDA, coordinates_t.device().index());
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:842]:
```suggestion
  auto d_options = torch::dtype(torch::kUInt8).device(coordinates_t.device());
```

File: torchani/cuaev/aev.cu [hunk 20]
```diff
@@ -571,14 +948,235 @@ torch::Tensor cuComputeAEV(
...
+
+  PairDist<float>* d_Rij = (PairDist<float>*)tensor_Rij.data_ptr();
+  PairDist<float>* d_radialRij = (PairDist<float>*)tensor_radialRij.data_ptr();
+  PairDist<float>* d_angularRij = (PairDist<float>*)tensor_angularRij.data_ptr();
+  PairDist<float>* d_centralAtom = (PairDist<float>*)tensor_centralAtom.data_ptr();
+  int* d_numPairsPerCenterAtom = (int*)tensor_numPairsPerCenterAtom.data_ptr();
+  int* d_centerAtomStartIdx = (int*)tensor_centerAtomStartIdx.data_ptr();
+
+  Tensor grad_radial_dist = torch::zeros(nRadialRij, coordinates_t.options().requires_grad(false));
+  // float* grad_radial_dist_ptr = (float*)grad_radial_dist.data_ptr();
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:1011]: Remove unused code

File: torchani/cuaev/aev.cu [hunk 21]
```diff
@@ -571,14 +948,235 @@ torch::Tensor cuComputeAEV(
...
+  nblocks = (nRadialRij + block_size - 1) / block_size;
+  pairwiseDistance_backward<<<nblocks, block_size, 0, stream>>>(
+      coordinates_t.packed_accessor32<float, 3, torch::RestrictPtrTraits>(),
+      grad_radial_dist.packed_accessor32<float, 1, torch::RestrictPtrTraits>(),
+      grad_coord.packed_accessor32<float, 3, torch::RestrictPtrTraits>(),
+      d_radialRij,
+      nRadialRij);
+
+  auto smem_size = [&aev_params](int max_nbrs, int ncatom_per_tpb) {
+    // int sm_aev = sizeof(float) * align<4>(aev_params.angular_length); // (angular_length / 4 + 1) * 4
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:1035]: Remove unused code

File: torchani/cuaev/aev.cu [hunk 22]
```diff
@@ -571,14 +948,235 @@ torch::Tensor cuComputeAEV(
...
+  return grad_coord;
+}
+
+#define AEV_INPUT                                                                                                   \
+  const Tensor &coordinates_t, const Tensor &species_t, double Rcr_, double Rca_, const Tensor &EtaR_t,             \
+      const Tensor &ShfR_t, const Tensor &EtaA_t, const Tensor &Zeta_t, const Tensor &ShfA_t, const Tensor &ShfZ_t, \
+      int64_t num_species_
+
+Tensor cuaev_cuda(AEV_INPUT) {
+  // printf("calling cuda \n");
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:1080]: Remove debugging print

File: tests/test_cuaev.py [hunk 23]
```diff
@@ -83,6 +110,25 @@ def testTripeptideMD(self):
                 _, cu_aev = self.cuaev_computer((species, coordinates))
                 self.assertEqual(cu_aev, aev)
 
+    def testTripeptideMDBackward(self):
+        for i in range(100):
+            datafile = os.path.join(path, 'test_data/tripeptide-md/{}.dat'.format(i))
+            with open(datafile, 'rb') as f:
+                coordinates, species, _, _, _, _, _, _ = pickle.load(f)
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:117]:
```suggestion
                coordinates, species, *_ = pickle.load(f)
```

File: tests/test_cuaev.py [hunk 24]
```diff
@@ -106,6 +171,28 @@ def testVeryDenseMolecule(self):
                 _, cu_aev = self.cuaev_computer((species, coordinates))
                 self.assertEqual(cu_aev, aev, atol=5e-5, rtol=5e-5)
 
+    def testVeryDenseMoleculeBackward(self):
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:174]: Leave a comment on why we need this:
```python
# Test big neighbor list size
```

File: tests/test_cuaev.py [hunk 25]
```diff
@@ -106,6 +171,28 @@ def testVeryDenseMolecule(self):
...
+        for i in range(100):
+            datafile = os.path.join(path, 'test_data/tripeptide-md/{}.dat'.format(i))
+            with open(datafile, 'rb') as f:
+                coordinates, species, _, _, _, _, _, _ = pickle.load(f)
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:178]:
```suggestion
                coordinates, species, *_ = pickle.load(f)
```

**Review by zasdfgbnm — APPROVED** [review]
Mostly good

**Pull request merged**